From 0fac3a9a884aee1ec9466cede46583c7d1a1f2f0 Mon Sep 17 00:00:00 2001
From: Tommy Lin <lin.tommysc@inventec.com>
Date: Wed, 9 Mar 2022 05:43:04 +0000
Subject: [PATCH 5/7] Modified sel file location path and add RecordID property

- Changed sel log path to "/var/lib/sel_log"
- Add DBUS property "RecordID" to check the current record index

Resolve rsyslog droping issue

Symptom/Reason:
    - Add SEL entry failed after bmc just boot up

Root Cause:
    - rsync is waiting for eth0 to ready
    - SEL record ID generation caould be time consuming due to file I/O
    - rsyslog droping log due to rate-limiting

Solution/Change:
    [systemd]
      - Add interface option for systemd-networkd-wait-online.service to wait on
    [phosphor-sel-logger]
      - Refine record ID generation procedure
    [rsyslog]
      - Increase RateLimit.Burst from 2000 to 20000

Entry Test:
    - ipmitool raw 0x0a 0x44 0x00 0x00 0x02 0x00 0x00 0x00 0x00 0x00 0x00 0x04 0x01 0x17 0x00 0xa0 0x04 0x07
---
 include/sel_logger.hpp |  3 ++-
 src/sel_logger.cpp     | 21 ++++++++++++++++++++-
 2 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/include/sel_logger.hpp b/include/sel_logger.hpp
index adf4b18..313e4e2 100644
--- a/include/sel_logger.hpp
+++ b/include/sel_logger.hpp
@@ -40,7 +40,8 @@ static constexpr size_t selEvtDataMaxSize = 3;
 static constexpr size_t selOemDataMaxSize = 13;
 static constexpr uint8_t selEvtDataUnspecified = 0xFF;
 
-static const std::filesystem::path selLogDir = "/var/log";
+unsigned int recordIndex = 0;
+static const std::filesystem::path selLogDir = "/var/lib/sel_log";
 static const std::string selLogFilename = "ipmi_sel";
 
 template <typename... T>
diff --git a/src/sel_logger.cpp b/src/sel_logger.cpp
index 30733aa..0463074 100644
--- a/src/sel_logger.cpp
+++ b/src/sel_logger.cpp
@@ -200,7 +200,7 @@ void clearSelLogFiles()
 static unsigned int getNewRecordId(void)
 {
 #ifndef SEL_LOGGER_CLEARS_SEL
-    static unsigned int recordId = initializeRecordId();
+    static unsigned int recordId = recordIndex;
 
     // If the log has been cleared, also clear the current ID
     std::vector<std::filesystem::path> selLogFiles;
@@ -214,6 +214,7 @@ static unsigned int getNewRecordId(void)
     {
         recordId = 1;
     }
+    recordIndex = recordId;
     return recordId;
 }
 #endif
@@ -369,6 +370,24 @@ int main(int, char*[])
             return selAddOemRecord(message, selData, recordType);
         });
 
+        // Add RecordID property to check the current record index
+        ifaceAddSel->register_property("RecordID", recordIndex,
+                                    // set handle
+                                    [&](const unsigned int& record,
+                                        unsigned int& propertyValue) {
+                                        if (record < selInvalidRecID)
+                                        {
+                                            propertyValue = record;
+                                        }
+
+                                        return 1;
+                                    },
+                                    // get handle
+                                    [&](unsigned int& propertyValue) {
+                                        propertyValue = recordIndex;
+                                        return propertyValue;
+                                    });
+
         ifaceAddSel->register_property("SEL_Enable", SEL_Enable,
                                     // set handle
                                     [&](const bool& status,
-- 
2.33.0

